### PR TITLE
Add a retry mechanism on all API calls

### LINKED ISSUE #314: Marge fails too hard on network errors
When margebot encounters a network error during the merging process (e.g., timeout when checking the CI status), it fails hard and adds a "I'm broken inside" comment. Because network errors can be transient, margebot should retry failed network requests.

I can try to make a PR for this, but I have the following questions:

- Do you agree with the diagnostic and with the main proposal?

- I'm not sure where to make the necessary modifications in the code. Specifically, I'm not sure what granularity to have:
    - At the high-level of `single_merge_job`'s `execute`: we won't miss any errors (at least not during the merging process)?
    - At a lower level (`fetch_approvals` and `update_merge_request_and_accept` or even lower): we have more specific context for what stage failed?
    - At a higher level (outside of `single_merge_job`)?
    - At a different level altogether such as by patching the `self._api` object or changing some configuration of the underlying http request library.

### PR DESCRIPTION
**Pull request by slovdahl**
fixes smarkets/marge-bot#314

Originally submitted by @z8v in https://github.com/smarkets/marge-bot/pull/349. Will keep it in our fork until it has been merged upstream.

### CODE REVIEW

File: tests/test_gitlab.py [hunk 1]
```diff
@@ -11,3 +19,23 @@ def test_parse_no_edition(self):
...
+        api = gitlab.Api(HTTPBIN, "", append_api_version=False)
+
+        with self.assertRaises(gitlab.Conflict):
+            api.call(GET("/status/409"))
+
+        with self.assertRaises(gitlab.TooManyRequests):
+            api.call(GET("/status/429"))
+
+        with self.assertRaises(gitlab.GatewayTimeout):
+            api.call(GET("/status/504"))
```
Comment (perlun) [inline tests/test_gitlab.py:41]: Side note, but we should enable GitHub Actions also now to ensure stuff like this gets tested. I looked at them and they are safe apart from the `deploy` workflow which I created a PR to remove now: https://github.com/hiboxsystems/marge-bot/pull/7

Comment (perlun) [reply tests/test_gitlab.py:41]: @slovdahl - if you amend the commit here and push it once more, the CI job should run. It'll likely fail just like in https://github.com/hiboxsystems/marge-bot/actions/runs/3747379517/jobs/6363544706 but I can give that a look in a separate PR.

**Review by perlun — APPROVED** [review]

File: tests/test_gitlab.py (hunk 1 above)
Comment (slovdahl) [reply tests/test_gitlab.py:41]: Done :heavy_check_mark:

**Review by perlun — APPROVED** [review]
Feel free to bypass CI failures for now. We'll get them fixed in a moment.

File: requirements.txt [hunk 2]
```diff
@@ -2,6 +2,7 @@ ConfigArgParse
 maya
 PyYAML
 requests
+retry
```
Comment (perlun) [inline requirements.txt]: This seems to be incomplete, I think it should somehow go into `requirements_frozen.txt` as well. Fails like this in CI: https://github.com/hiboxsystems/marge-bot/actions/runs/3747611909/jobs/6364021290#step:5:165

Comment (slovdahl) [reply requirements.txt]: Good point, added it and an entry to `requirements_frozen.txt` too. :drum: :crossed_fingers:

Comment (slovdahl) [reply requirements.txt]: Rebased on top of master and added `retry2` as a Poetry dependency instead.

Comment (slovdahl) [reply requirements.txt]: The test is still not working though.. So need to look into that before this can be merged.

**Comment by xtermi2** [comment]
Is here any progress? We need this retry mechanism because he have a lot of failures which could be solved by retries 🙏

**Comment by slovdahl** [comment]
Hi @xtermi2! This PR has stalled a little bit, we have not had any problems recently that would be solved by this. But I do agree it's a nice improvement in general.

Overall, we're in the process of consolidating all the current and active forks of `marge-bot` into one: https://gitlab.com/marge-org/marge-bot. So right now it's unlikely that this fork will see much future development.

**Pull request merged**